pkg/clusteragent/clusterchecks/dispatcher_ksm.go

```diff
@@ -0,0 +1,183 @@
+	d.ksmShardingMutex.Lock()
+	defer d.ksmShardingMutex.Unlock()
+
+	d.ksmShardedConfig = config
+	d.ksmShardedDigests = shardedDigests
+}
+
+// unscheduleKSMCheck removes all sharded KSM configs if this is a sharded KSM check
+// Returns true if sharded configs were removed, false otherwise
+// Thread-safe: protected by ksmShardingMutex
+func (d *dispatcher) unscheduleKSMCheck(config integration.Config) bool {
+	if !d.ksmSharding.IsEnabled() || !d.ksmSharding.IsKSMCheck(config) {
+		return false
+	}
+
+	if !d.isAlreadySharded(config) {
+		return false
+	}
+
+	// Atomically read the digests and clear tracking
+	// We must hold the lock to avoid TOCTOU race between checking and reading digests
+	d.ksmShardingMutex.Lock()
+	digestsCopy := make([]string, len(d.ksmShardedDigests))
+	copy(digestsCopy, d.ksmShardedDigests)
+	digestCount := len(digestsCopy)
+
+	// Clear the tracking while holding the lock
+	d.ksmShardedConfig = integration.Config{}
+	d.ksmShardedDigests = nil
+	d.ksmShardingMutex.Unlock()
+
+	log.Infof("Unscheduling sharded KSM check %s (removing %d shards)", config.Digest(), digestCount)
+
+	// Remove all sharded configs (outside the lock since this is expensive)
+	for _, digest := range digestsCopy {
+		log.Debugf("Removing KSM shard with digest %s", digest)
+		d.removeConfig(digest)
+	}
+
+	log.Infof("Successfully unscheduled all sharded KSM configs")
+	return true
+}
+
+// validateClusterTaggerForKSM validates that cluster tagger is enabled for KSM resource sharding.
+// Returns true if Kubernetes resource collection is enabled, false if disabled.
+// Logs warnings if disabled, informing users about potential tag limitations.
+// Sharding always proceeds regardless of return value - this only informs the caller of the status.
+func (d *dispatcher) validateClusterTaggerForKSM() bool {
+	// KSM uses the cluster tagger only for namespace labels/annotations as tags.
+	// When cluster_agent.collect_kubernetes_tags is enabled:
+	// - DCA collects namespace objects into workloadmeta
+	// - Cluster tagger can propagate namespace labels/annotations as tags to objects in that namespace
+	//
+	// Note: This only works with global tag configuration (e.g., kubernetesResourcesLabelsAsTags),
+	// not with check-specific label configuration.
+	//
+	// KSM does NOT use the cluster tagger for pod→deployment or other cross-resource relationships.
+	// Those relationships are handled by KSM itself through the Kubernetes API.
+	//
+	// Important: Sharding can break the label_joins option in KSM configs, because label_joins
+	// requires the related objects to be collected in the same KSM instance. If your config uses
+	// label_joins across different resource types (e.g., joining pod labels to node metrics),
+	// those resources must be in the same shard or label_joins will not work correctly.
+
+	collectKubernetesTags := pkgconfigsetup.Datadog().GetBool("cluster_agent.collect_kubernetes_tags")
```

Comment:
This function is not correct. `cluster_agent.collect_kubernetes_tags` is an option to explicitly enable pod collection on the DCA. I know the name is confusing because it seems to imply that it's global, but you can check that it's only related to pods.

Even when `cluster_agent.collect_kubernetes_tags` is false, if namespace labels/annotations as tags are specified in the global option, the cluster tagger will collect the necessary namespace tags and this will work.